src/Features/Core/Portable/Diagnostics/Analyzers/NamingStyles/Serialization/SymbolSpecification.cs

```diff
@@ -228,9 +229,35 @@ internal static SymbolKindOrTypeKind AddTypeKindFromXElement(XElement typeKindEl
             {
                 return new SymbolKindOrTypeKind((TypeKind)Enum.Parse(typeof(TypeKind), typeKindElement.Value));
             }
+
+            public override bool Equals(object obj)
+            {
+                return Equals(obj as SymbolKindOrTypeKind);
+            }
+
+            public bool Equals(SymbolKindOrTypeKind other)
+            {
+                return other != null &&
+                       this.SymbolKind == other.SymbolKind &&
+                       this.TypeKind == other.TypeKind;
+            }
+
+            public override int GetHashCode()
+            {
+                if (this.SymbolKind.HasValue && this.TypeKind.HasValue)
+                {
+                    return Hash.Combine(this.SymbolKind.Value.GetHashCode(), this.TypeKind.Value.GetHashCode());
+                }
+                else if(this.SymbolKind.HasValue)
+                {
+                    return this.SymbolKind.Value.GetHashCode();
+                }
+
+                return this.TypeKind.Value.GetHashCode();
+            }
```

Comment:
Consider also providing `operator==` and `operator!=` anytime you override `object.Equals`